src/expr.rs

```diff
@@ -594,27 +604,24 @@ fn rewrite_block_inner(
     allow_single_line: bool,
     context: &RewriteContext<'_>,
     shape: Shape,
-) -> Option<String> {
-    let prefix = block_prefix(context, block, shape)?;
+) -> RewriteResult {
+    let prefix = block_prefix(context, block, shape).unknown_error()?;
 
     // shape.width is used only for the single line case: either the empty block `{}`,
     // or an unsafe expression `unsafe { e }`.
-    if let rw @ Some(_) = rewrite_empty_block(context, block, attrs, label, &prefix, shape) {
-        return rw;
+    if let Some(rw_str) = rewrite_empty_block(context, block, attrs, label, &prefix, shape) {
+        return Ok(rw_str);
     }
 
-    let result = rewrite_block_with_visitor(context, &prefix, block, attrs, label, shape, true);
-    if let Some(ref result_str) = result {
-        if allow_single_line && result_str.lines().count() <= 3 {
-            if let rw @ Some(_) =
-                rewrite_single_line_block(context, &prefix, block, attrs, label, shape)
-            {
-                return rw;
-            }
+    let result_str =
+        rewrite_block_with_visitor(context, &prefix, block, attrs, label, shape, true)?;
+    if allow_single_line && result_str.lines().count() <= 3 {
+        if let rw @ Ok(_) = rewrite_single_line_block(context, &prefix, block, attrs, label, shape)
+        {
+            return rw;
         }
     }
-
-    result
+    Ok(result_str)
```

Comment:
This doesn't look like a one-to-one translations from `Option<String>` -> `RewriteResult`. Also, doesn't `rewrite_block_with_visitor` now return `RewriteResult`? I would have expected to see something closer this:

```rust
    let result = rewrite_block_with_visitor(context, &prefix, block, attrs, label, shape, true);
    if let Ok(ref result_str) = result {
        if allow_single_line && result_str.lines().count() <= 3 {
            if let rw @ Ok(_) = rewrite_single_line_block(context, &prefix, block, attrs, label, shape)
            {
                return rw;
            }
        }
    }
```

Reply:
Yes, it isn't a one-to-one translation but I made this change intentionally. 

From what I understood of the original code:
```
(1) If rewrite_block_with_visitor fails, return None.
(2) If rewrite_block_with_visitor succeeds, check whether we should try to format it into a single-line block.
      a. If rewrite_single_line_block succeeds, return that result.
      b. If rewrite_single_line_block fails, just return the result from rewrite_block_with_visitor.
```
        
My intention was to simplify `if let Ok(ref result_str) = result {...}` with using `?` operator on calling `rewrite_block_with_visitor`. I thought it won't make a difference in control flow or formatting logic described above. 
Still, if you think this change does not worth the change or it actually breaks the original control flow, please tell me.

Reply:
After taking another look at this the change seems to preserve the original behavior so we're good here!